src/objective-c/GRPCClient/GRPCCall.m

```diff
@@ -139,8 +146,10 @@ - (void)finishWithError:(NSError *)errorOrNil {
   _self = nil;
 
   // If there were still request messages coming, stop them.
-  _requestWriter.state = GRXWriterStateFinished;
-  _requestWriter = nil;
+  @synchronized(_requestWriter) {
```

Comment:
> you should never **@synchronize**(nil)

I was going by this: http://rykap.com/objective-c/2015/05/09/synchronized.html. The guy tested that `@synchronized(nil)` just bypasses locking; and setting the synchronized variable to `nil` inside the block is ok (in the sense that it doesn't prevent the lock from unlocking at the end of the block). But on further reading, apparently it's undocumented by Apple?

Synchronizing on `self` wouldn't stop the code from messaging `_requestWriter` concurrently with other threads outside my control (the requests writer is passed by the user on `init...`); so it really needs to be `@synchronized(_requestWriter)`.

I'm thinking a way out of this is just to _not set_ it to `nil`. When the writer finishes (because we tell it so, or because it run out of stuff to write), it'll release its reference to us (the call object is its writeable), so the retain cycle writer <-> call is guaranteed to be broken.